atc/steps.go

```diff
@@ -46,42 +47,52 @@ func (step *Step) UnmarshalJSON(data []byte) error {
 		return err
 	}
 
-	var outerStep StepConfig
+	var prevStep StepWrapper
 	for _, s := range StepPrecedence {
 		_, found := deferred[s.Key]
 		if !found {
 			continue
 		}
 
-		step := s.New()
+		curStep := s.New()
 
-		err := step.ParseJSON(data)
+		err := json.Unmarshal(data, curStep)
 		if err != nil {
 			return MalformedStepError{
 				StepType: s.Key,
 				Err:      err,
 			}
 		}
 
-		if outerStep == nil {
-			outerStep = step
-		} else {
-			outerStep.Wrap(step)
+		if step.Config == nil {
+			step.Config = curStep
+		}
+
+		if prevStep != nil {
+			prevStep.Wrap(curStep)
 		}
 
-		delete(deferred, s.Key)
+		deleteKeys(deferred, curStep)
+
+		if wrapper, isWrapper := curStep.(StepWrapper); isWrapper {
+			prevStep = wrapper
+		} else {
+			break
+		}
 
 		data, err = json.Marshal(deferred)
 		if err != nil {
 			return fmt.Errorf("re-marshal deferred parsing: %w", err)
 		}
 	}
 
-	if outerStep == nil {
+	if step.Config == nil {
 		return ErrNoStepConfigured
 	}
 
-	step.Config = outerStep
+	if len(deferred) != 0 {
+		step.UnknownFields = deferred
```

Comment:
Hmm yeah i agree that it’s kind of weird. We started out by trying to return a public error type (UnknownFieldsError) that we could special case on to ignore when we don’t care about unknown fields (whenever we pull from the database).

The problem we ran into was that we unmarshal a full atc.JobConfig from the database - so even if we choose to ignore the error, unmarshalling the rest of the JobConfig would short-circuit.

I’d be happy to hear any other ideas you might have. You also mentioned in the issue that you were looking at a yaml library that could help with this - I’m not sure exactly how it’d help, but if you think that’s a viable option we could look into that

Reply:
Another possibility - unmarshal it normally (in a non-strict way) without keeping track of unknown fields. When we save the pipeline, we do the following:

1. Unmarshal the config bytes into an `atc.Config` as we do now
2. Unmarshal the config bytes into a `map[string]interface{}`
3. Marshal the `atc.Config` again
4. Unmarshal those bytes into a `map[string]interface{}`
5. Traverse the `map`s and figure out which fields are missing in the second `map`

What do you think?

Reply:
Though, one nice thing about storing the UnknownFields is that we can remarshal it and preserve those fields. That way, when we compute the diff in a fly set-pipeline after removing the invalid fields, it’ll properly show that it’s being removed (and otherwise, we’d need something hacky to avoid fly complaining that there are no changes, since the existing config we got back from the API wouldn’t contain the invalid fields)

Reply:
We can in the future decide to become stricter and raise such errors in `UnmarshalJSON` rather than through the validator, without worrying about the consumer ( fetch from DB vs. set ), after what we consider `sufficient` time and we can think of storing the unknown fields as a temporary solution to address deprecation

Reply:
@aoldershaw 

> You also mentioned in the issue that you were looking at a yaml library that could help with this - I’m not sure exactly how it’d help, but if you think that’s a viable option we could look into that

This ended up being a bit of a dead end. The idea I was exploring was to eliminate the multi-pass unmarshalling by making use of `yaml:",inline"` tags instead, like this:

```go
type TimeoutStep struct {
    Parent Step `yaml:",inline"`
    Duration string `yaml:"timeout"`
}
```

Unfortunately both `goccy/go-yaml` and `gopkg.in/yaml.v3` (we currently use v2, but v3 allows inline structs to use `UnmarshalYAML`) both just go into an infinite loop as the `UnmarshalYAML` function called on the inline `Step` is called with *all* fields, including the ones that were already "parsed" (the `timeout:` field). I was able to modify `goccy/go-yaml` to avoid this by calling it with only the remaining fields, but I'm still not super confident in switching YAML parsers. I'm considering making the same change to `gopkg.in/yaml.v3` but to be honest I'm running out of steam on this whole side venture.

The whole point of this was to try and get everything parsed in one pass so that the original strict-or-not nature of the parsing would be respected, but I think even that would require further changes to `gopkg.in/yaml.v3` such that the `node.Decode` call respects the original `knownFields` config that constructed the node. I believe right now it just does a non-strict decode.

---

re: https://github.com/concourse/concourse/pull/5878#discussion_r454308105 - that would work, but eh, it might not be worth it. I think at this point I'm OK with just leaving this as-is. Another option to consider may be [`mapstructure`](https://godoc.org/github.com/mitchellh/mapstructure), but it makes things a little more complicated (we used to use it in the past and switched *from* it).

I do see some merit in handling this entirely through validation though and just never doing strict unmarshalling. Heck, maybe an approach like you suggested wouldn't be too crazy.

Here's one thing that could improve: we allow unknown keys at the toplevel, which works through this funny dance:

https://github.com/concourse/concourse/blob/edf93993cb7406f700d1147eea214054d8c2d843/atc/config.go#L28-L36

If we were to implement this entirely through validation instead, we could enforce that policy in a less awkward way. 🤷‍♂️ But again, I'm OK with this current approach, at least for getting 6.4.0 out.

Reply:
@vito Cool, thanks for the details! We looked into using mapstructure, but yeah, it seemed like it’d be more complicated than was worth it for our use case.

I agree we should get this out for 6.4.0, but here’s a (possbly foolish) idea for the future - what if we had a migration that removed any existing unknown fields from all job configs, and then added back strict unmarshaling. It may be unexpected for these fields to be removed on migration, but they weren’t actually doing anything 🤷‍♂️

Reply:
@aoldershaw Migrations that modify config are pretty painful to write: they have to decrypt and then unmarshal into copied type definitions<sup>1</sup> and then write the updated config back all while only using one transaction. In the past this had led to migrations which try to hold everything in memory which inevitably fails on large clusters.

I wouldn't recommend it. 😅

<sup>1</sup> So they don't accidentally 'stray forward' when the `atc/` package changes.